extensions/standard-processors/processors/LogAttribute.h

```diff
@@ -90,26 +90,19 @@ class LogAttribute : public core::Processor {
   class ReadCallback : public InputStreamCallback {
    public:
     ReadCallback(uint64_t size)
-        : read_size_(0) {
-      buffer_size_ = size;
-      buffer_ = new uint8_t[buffer_size_];
-    }
-    ~ReadCallback() {
-      if (buffer_)
-        delete[] buffer_;
+        : buffer_(size)  {
     }
     int64_t process(std::shared_ptr<io::BaseStream> stream) {
-      int64_t ret = 0;
-      ret = stream->read(buffer_, buffer_size_);
-      if (!stream)
-        read_size_ = stream->getSize();
-      else
-        read_size_ = buffer_size_;
-      return ret;
+      if (buffer_.size() == 0U) {
+        return 0U;
+      }
+      int ret = stream->read(buffer_.data(), buffer_.size());
```

Comment:
Am I misunderstanding something? Less data is read from the stream therefore we throw an error? The contract for the streams, I believe, returns the size read and the input size isn't reflective of an error condition.

Reply:
@phrocker To be honest, I am not sure what the contract really is in these cases. I could not find relevant comments or documentation, so I am basing my understanding on the usages I have seen and the implementation of ContentRepositories, ResourceClaim and BaseStream.
What this implementation did before I changed it was the following: it requested the size of the FlowFile and created a ReadCallback with a buffer size equal to the FlowFile's size.
This ReadCallback is then passed to ProcessSession::read which gets a BaseStream from the ContentRepository for the ResourceClaim owned by the FlowFile. It then passes this BaseStream to the ReadCallback's process function where we are currently.
The old implementation then read at most buffer_size_ (so the FlowFile's size) bytes from this stream.
**It then checked whether the stream (which is a shared_ptr) was null, right after it dereferenced it and called a function on it!**
If this code path could have ever been reached (which it couldn't have been, because we would have gotten a SIGSEGV before that), it would have set the read_size_ to the size of the stream, ignoring the return value of the read operation on the stream.
Otherwise (so in every case), it just sets the read_size_ to buffer_size_ (again, ignoring the size returned by read), which is, in fact, the FlowFile's size.
So this ReadCallback could only ever have worked, if we could indeed read all data in one read operation from the stream provided by the ContentRepository. And this is what I think the contract is, and this is how I have seen it used throughout the code. (We assume that in this case read will return all data, and not just partial data).
My implementation just cleans up this logic adds an extra check to see whether this contract is really held true.

Reply:
I don't think that's a correct interpretation of the contract. The null dereference is bad, but my focus is continuation of the same positive path behavior. My opinion is that this should be chunking and checking return size. The fact that the constructor accepts the size doesn't make a lot of sense. We should choose a smaller size and chunk it ( which the contract does support even if it isn't evident ) or be safe and continue reading until -1 or 0 are met.

-1 returning rolls back, but when that exception is caught the same will be done so the change isn't great. The only change is that a partial read ( which the contract doesn't prohibit ). I'll have to think about this general change and the impact and realistic possibility it may occur.

Reply:
Let's leave my comment at this since this is what I mostly care about: I don't mind the change in behavior ( exhibited or theoretical ) -- but can you add a log statement to show how many bytes were read vs requested? 

This processor is used chiefly as a debug tool. Notwithstanding the issues you resolved, if the number of bytes read != stream, I think the user would want to know about that and it may help us triage a potential problem in a repo or on disk. That would mean the flow file size does not match the content size -- and I don't know if that would come across very well even with the code that exists in master.

Reply:
I agree with you that chucking is generally a better solution for this, if for nothing else than for the lower memory requirement. In this case, because we are only dealing with streams no larger than 1 MB, this benefit is not that pronounced.

I think adding the log you mentioned is a good idea, I don't know why I didn't do it, I will do now.

Just to make sure that I have correct understanding: we expect that we will be able to read exactly as many bytes from the stream as flow_file->getSize() returns (let it be in one piece or in chunks). And if that is not the case, it means there is a bug somewhere.